File: apps/api/src/caseops_api/schemas/ip_specialist.py
"""Independent specialist intake facts; no inferred rights or legal deadlines."""

from __future__ import annotations

from datetime import date, datetime
from typing import Annotated, Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, StringConstraints

Text = Annotated[
    str,
    StringConstraints(strip_whitespace=True, min_length=1, max_length=500, pattern=r"^[^\x00]*$"),
]
Narrative = Annotated[
    str,
    StringConstraints(strip_whitespace=True, min_length=1, max_length=4000, pattern=r"^[^\x00]*$"),
]
Domain = Literal[
    "design",
    "copyright",
    "domain_name",
    "licensing",
    "enforcement",
    "geographical_indication",
    "plant_variety",
    "semiconductor_layout",
    "trade_secret",
    "customs_enforcement",
]


class StrictModel(BaseModel):
    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)


class DesignFacts(StrictModel):
    domain: Literal["design"] = "design"
    applicant: Text
    article: Text
    classification_as_supplied: Text | None = None
    novelty_statement: Narrative | None = None
    representation_description: Narrative | None = None
    publication_instruction: Literal["unknown", "confidential", "published"] = "unknown"


class CopyrightFacts(StrictModel):
    domain: Literal["copyright"] = "copyright"
    work_type_as_supplied: Text
    author: Text
    claimant: Text
    ownership_claim: Narrative
    created_on: date | None = None
    published_on: date | None = None
    ownership_disputed: bool = False


class DomainNameFacts(StrictModel):
    domain: Literal["domain_name"] = "domain_name"
    domain_name: Text
    registrar: Text | None = None
    registrant_as_supplied: Text
    expiry_as_supplied: date | None = None
    dispute_channel_as_supplied: Text | None = None


class LicensingFacts(StrictModel):
    domain: Literal["licensing"] = "licensing"
    grantor: Text
    grantee: Text
    affected_rights_as_supplied: Narrative
    territory: Text
    field_of_use: Text | None = None
    exclusivity: Literal["not_reviewed", "exclusive", "non_exclusive", "sole"] = "not_reviewed"
    term_as_supplied: Text | None = None
    royalty_terms_confidential: Narrative | None = None
    interpretation_issue: Narrative | None = None


class EnforcementFacts(StrictModel):
    domain: Literal["enforcement"] = "enforcement"
    represented_party: Text
    asserted_right_as_supplied: Narrative
    allegation: Narrative
    proposed_channel: Literal[
        "investigation", "notice", "platform", "customs", "opposition", "cancellation", "litigation"
    ]
    alleged_party: Text | None = None


class GeographicalIndicationFacts(StrictModel):
    domain: Literal["geographical_indication"] = "geographical_indication"
    applicant_or_association: Text
    geographical_area: Narrative
    goods: Text
    specification_as_supplied: Narrative | None = None
    authorised_user_claim: Text | None = None


class PlantVarietyFacts(StrictModel):
    domain: Literal["plant_variety"] = "plant_variety"
    denomination: Text
    category_as_supplied: Text
    crop_species: Text
    applicant: Text
    breeder_as_supplied: Text | None = None
    farmer_claim_as_supplied: Text | None = None
    material_custodian: Text
    material_access_instructions: Narrative


class SemiconductorLayoutFacts(StrictModel):
    domain: Literal["semiconductor_layout"] = "semiconductor_layout"
    creator: Text
    proprietor_as_supplied: Text
    layout_description: Narrative
    first_commercial_exploitation: date | None = None
    exploitation_territory: Text | None = None


class TradeSecretFacts(StrictModel):
    domain: Literal["trade_secret"] = "trade_secret"
    owner_as_supplied: Text
    custodian: Text
    asset_reference: Text
    protective_controls: Narrative
    permitted_access_as_supplied: Narrative
    # The register stores references and controls, never the secret substance.


class CustomsFacts(StrictModel):
    domain: Literal["customs_enforcement"] = "customs_enforcement"
    right_holder: Text
    asserted_right_as_supplied: Narrative
    products: Narrative
    authority_or_channel_as_supplied: Text
    authentication_guide_custodian: Text
    authorised_importer_policy: Narrative
    instruction_reference: Text


DomainFacts = Annotated[
    DesignFacts
    | CopyrightFacts
    | DomainNameFacts
    | LicensingFacts
    | EnforcementFacts
    | GeographicalIndicationFacts
    | PlantVarietyFacts
    | SemiconductorLayoutFacts
    | TradeSecretFacts
    | CustomsFacts,
    Field(discriminator="domain"),
]


class SpecialistFacts(StrictModel):
    title: Annotated[
        str,
        StringConstraints(
            strip_whitespace=True, min_length=1, max_length=255, pattern=r"^[^\x00]*$"
        ),
    ]
    client_id: UUID
    jurisdiction_as_supplied: Text
    details: DomainFacts


class SpecialistCorrection(StrictModel):
    expected_version: int = Field(ge=1)
    expected_lifecycle_version: int = Field(ge=0)
    reason: Text
    facts: SpecialistFacts


class SpecialistSource(StrictModel):
    document_id: UUID
    document_version_id: UUID
    content_sha256: str = Field(pattern=r"^[a-f0-9]{64}$")
    locator: Text


class SpecialistObservation(StrictModel):
    expected_version: int = Field(ge=1)
    expected_lifecycle_version: int = Field(ge=0)
    kind: Annotated[str, StringConstraints(min_length=1, max_length=80)]
    occurred_on: date
    account: Narrative
    source: SpecialistSource
    supersedes_id: UUID | None = None


class SpecialistRecord(StrictModel):
    id: UUID
    docket_id: UUID
    asset_id: UUID
    contract_version: str
    version: int
    lifecycle_version: int
    lifecycle_status: str
    is_active: bool
    created_at: datetime
    facts: SpecialistFacts


class SpecialistSummary(StrictModel):
    id: UUID
    domain: Domain
    title: str
    lifecycle_status: str
    version: int


class SpecialistList(StrictModel):
    records: list[SpecialistSummary]
    next_cursor: UUID | None


class SpecialistObservationRecord(StrictModel):
    id: UUID
    sequence: int
    kind: str
    occurred_on: date
    account: str
    source: SpecialistSource
    supersedes_id: UUID | None
    recorded_at: datetime
    legal_effect: Literal["not_determined"] = "not_determined"


class SpecialistObservationList(StrictModel):
    observations: list[SpecialistObservationRecord]
    next_cursor: int | None


class SpecialistField(StrictModel):
    key: str
    label: str
    kind: Literal["text", "textarea", "date", "boolean", "select"]
    required: bool
    max_length: int | None = None
    options: list[str] = Field(default_factory=list)


class SpecialistContract(StrictModel):
    domain: Domain
    label: str
    contract_version: str
    contract_path: str
    child_prd_sha256: str
    intake_available: bool
    fields: list[SpecialistField]
    observation_kinds: list[str]
    blockers: list[str]

File: apps/web/lib/api/ip-specialist.ts
import { z } from "zod";

import { apiRequest } from "./client";

const text = (max = 500) => z.string().trim().min(1).max(max).regex(/^[^\0]*$/);
const optional = text().nullable().optional();
const narrative = text(4000);
const optionalNarrative = narrative.nullable().optional();
const optionalDate = z.string().date().nullable().optional();
const domainSchema = z.enum(["design", "copyright", "domain_name", "licensing", "enforcement",
  "geographical_indication", "plant_variety", "semiconductor_layout", "trade_secret", "customs_enforcement"]);
export const specialistDetailsSchema = z.discriminatedUnion("domain", [
  z.object({ domain: z.literal("design"), applicant: text(), article: text(), classification_as_supplied: optional,
    novelty_statement: optionalNarrative, representation_description: optionalNarrative,
    publication_instruction: z.enum(["unknown", "confidential", "published"]).default("unknown") }).strict(),
  z.object({ domain: z.literal("copyright"), work_type_as_supplied: text(), author: text(), claimant: text(),
    ownership_claim: narrative, created_on: optionalDate, published_on: optionalDate, ownership_disputed: z.boolean().default(false) }).strict(),
  z.object({ domain: z.literal("domain_name"), domain_name: text(), registrar: optional, registrant_as_supplied: text(),
    expiry_as_supplied: optionalDate, dispute_channel_as_supplied: optional }).strict(),
  z.object({ domain: z.literal("licensing"), grantor: text(), grantee: text(), affected_rights_as_supplied: narrative,
    territory: text(), field_of_use: optional, exclusivity: z.enum(["not_reviewed", "exclusive", "non_exclusive", "sole"]).default("not_reviewed"),
    term_as_supplied: optional, royalty_terms_confidential: optionalNarrative, interpretation_issue: optionalNarrative }).strict(),
  z.object({ domain: z.literal("enforcement"), represented_party: text(), asserted_right_as_supplied: narrative,
    allegation: narrative, proposed_channel: z.enum(["investigation", "notice", "platform", "customs", "opposition", "cancellation", "litigation"]),
    alleged_party: optional }).strict(),
  z.object({ domain: z.literal("geographical_indication"), applicant_or_association: text(), geographical_area: narrative,
    goods: text(), specification_as_supplied: optionalNarrative, authorised_user_claim: optional }).strict(),
  z.object({ domain: z.literal("plant_variety"), denomination: text(), category_as_supplied: text(), crop_species: text(),
    applicant: text(), breeder_as_supplied: optional, farmer_claim_as_supplied: optional, material_custodian: text(), material_access_instructions: narrative }).strict(),
  z.object({ domain: z.literal("semiconductor_layout"), creator: text(), proprietor_as_supplied: text(), layout_description: narrative,
    first_commercial_exploitation: optionalDate, exploitation_territory: optional }).strict(),
  z.object({ domain: z.literal("trade_secret"), owner_as_supplied: text(), custodian: text(), asset_reference: text(),
    protective_controls: narrative, permitted_access_as_supplied: narrative }).strict(),
  z.object({ domain: z.literal("customs_enforcement"), right_holder: text(), asserted_right_as_supplied: narrative, products: narrative,
    authority_or_channel_as_supplied: text(), authentication_guide_custodian: text(), authorised_importer_policy: narrative, instruction_reference: text() }).strict(),
]);
export const specialistFactsSchema = z.object({ title: text(255), client_id: z.string().uuid(),
  jurisdiction_as_supplied: text(), details: specialistDetailsSchema }).strict();
export const specialistRecordSchema = z.object({ id: z.string().uuid(), docket_id: z.string().uuid(), asset_id: z.string().uuid(),
  contract_version: z.string(), version: z.number().int().positive(), lifecycle_version: z.number().int().nonnegative(),
  lifecycle_status: z.string(), is_active: z.boolean(), created_at: z.string().datetime({ offset: true }), facts: specialistFactsSchema }).strict();
const specialistFieldSchema = z.object({ key: z.string(), label: z.string(), kind: z.enum(["text", "textarea", "date", "boolean", "select"]),
  required: z.boolean(), max_length: z.number().int().positive().nullable(), options: z.array(z.string()) }).strict();
const contractSchema = z.object({ domain: domainSchema, label: z.string(), contract_version: z.string(), contract_path: z.string(),
  child_prd_sha256: z.string().regex(/^[a-f0-9]{64}$/), intake_available: z.boolean(), fields: z.array(specialistFieldSchema),
  observation_kinds: z.array(z.string()), blockers: z.array(z.string()) }).strict();
const sourceSchema = z.object({ document_id: z.string().uuid(), document_version_id: z.string().uuid(),
  content_sha256: z.string().regex(/^[a-f0-9]{64}$/), locator: text() }).strict();
const observationSchema = z.object({ id: z.string().uuid(), sequence: z.number().int().positive(), kind: z.string(), occurred_on: z.string().date(),
  account: narrative, source: sourceSchema, supersedes_id: z.string().uuid().nullable(), recorded_at: z.string().datetime({ offset: true }),
  legal_effect: z.literal("not_determined") }).strict();
const listSchema = z.object({ records: z.array(z.object({ id: z.string().uuid(), domain: domainSchema, title: z.string(),
  lifecycle_status: z.string(), version: z.number().int().positive() }).strict()).max(50), next_cursor: z.string().uuid().nullable() }).strict();
const observationsSchema = z.object({ observations: z.array(observationSchema).max(50), next_cursor: z.number().int().positive().nullable() }).strict();

export type SpecialistFacts = z.infer<typeof specialistFactsSchema>;
export type SpecialistRecord = z.infer<typeof specialistRecordSchema>;
export type SpecialistContract = z.infer<typeof contractSchema>;
export type SpecialistObservation = z.infer<typeof observationSchema>;
const base = "/api/ip/specialist";
export const specialistKey = (id: string) => ["ip", "specialist-record", id];

export async function fetchSpecialistContracts(signal?: AbortSignal) {
  return z.array(contractSchema).max(10).parse(await apiRequest(`${base}/contracts`, { signal }));
}
export async function fetchSpecialistRecords(domain: string, closed: boolean, cursor?: string, signal?: AbortSignal) {
  const query = new URLSearchParams({ domain, include_closed: String(closed), limit: "25" });
  if (cursor) query.set("cursor", cursor);
  return listSchema.parse(await apiRequest(`${base}/records?${query}`, { signal }));
}
export async function fetchSpecialistRecord(id: string, version?: number, signal?: AbortSignal) {
  return specialistRecordSchema.parse(await apiRequest(`${base}/records/${encodeURIComponent(id)}${version ? `/versions/${version}` : ""}`, { signal }));
}
export async function saveSpecialistFacts(facts: SpecialistFacts, key: string, saved?: SpecialistRecord, reason?: string) {
  return specialistRecordSchema.parse(await apiRequest(`${base}/records${saved ? `/${saved.id}/corrections` : ""}`, {
    method: "POST", headers: { "Idempotency-Key": key }, body: saved ? { expected_version: saved.version,
      expected_lifecycle_version: saved.lifecycle_version, reason, facts: specialistFactsSchema.parse(facts) } : specialistFactsSchema.parse(facts),
  }));
}
export async function fetchSpecialistObservations(id: string, cursor?: number, signal?: AbortSignal) {
  const query = new URLSearchParams({ limit: "25" });
  if (cursor) query.set("cursor", String(cursor));
  return observationsSchema.parse(await apiRequest(`${base}/records/${id}/observations?${query}`, { signal }));
}
export async function saveSpecialistObservation(record: SpecialistRecord, input: Pick<SpecialistObservation, "kind" | "occurred_on" | "account" | "source" | "supersedes_id">, key: string) {
  return observationSchema.parse(await apiRequest(`${base}/records/${record.id}/observations`, { method: "POST",
    headers: { "Idempotency-Key": key }, body: { ...input, expected_version: record.version, expected_lifecycle_version: record.lifecycle_version } }));
}

export async function uploadSpecialistSource(record: SpecialistRecord, file: File, taxonomyKey: string) {
  const body = new FormData();
  body.append("upload", file);
  body.append("metadata_json", JSON.stringify({ taxonomy_key: taxonomyKey, title: file.name, confidentiality: "restricted",
    asset_type: record.facts.details.domain, links: [{ target_type: "docket", target_id: record.docket_id }] }));
  return apiRequest<{ outcome: string }>("/api/ip/documents/upload", { method: "POST", body });
}
